### PR TITLE
add ZeroCopy in transaction and block

### PR DESCRIPTION
**Pull request by lucas7788**
add ZeroCopy in transaction and block

### CODE REVIEW

File: common/uint160.go [hunk 1]
```diff
@@ -110,21 +109,21 @@ func ToScriptHash(address string) (Uint160, error) {
...
-	log.Tracef("[AddressToProgramHash] programhash: %x", ph.ToArray())
+	//log.Tracef("[AddressToProgramHash] programhash: %x", ph.ToArray())
 
 	addr, err := ph.ToAddress()
 	if err != nil {
 		return Uint160{}, err
 	}
 
-	log.Tracef("[AddressToProgramHash] address: %s", addr)
+	//log.Tracef("[AddressToProgramHash] address: %s", addr)
```
Comment (laizy) [inline common/uint160.go]: remove commented code

File: core/asset/asset.go [hunk 2]
```diff
@@ -65,6 +66,15 @@ func (a *Asset) Serialize(w io.Writer) error {
 	return nil
 }
 
+func (a *Asset)Serialization(sink *common.ZeroCopySink) error {
+	sink.WriteString(a.Name)
+	sink.WriteString(a.Description)
+	sink.WriteBytes([]byte{byte(a.Precision)})
+    sink.WriteBytes([]byte{byte(a.AssetType)})
+	sink.WriteBytes([]byte{byte(a.RecordType)})
```
Comment (laizy) [inline core/asset/asset.go]: use `sink.WriteByte(a.Precision)`

File: core/asset/asset.go [hunk 3]
```diff
@@ -98,3 +108,23 @@ func (a *Asset) Deserialize(r io.Reader) error {
 	}
 	return nil
 }
+
+func (a *Asset) Deserialization(source *common.ZeroCopySource) error {
+	name, _, irregular, eof := source.NextString()
+	description, _, irregular, eof := source.NextString()
+	p, _, irregular, eof := source.NextVarBytes()
+	t, _, irregular, eof := source.NextVarBytes()
```
Comment (laizy) [inline core/asset/asset.go]: test this function, it seems buggy.

File: core/asset/asset.go [hunk 4]
```diff
@@ -98,3 +108,23 @@ func (a *Asset) Deserialize(r io.Reader) error {
 	}
 	return nil
 }
+
+func (a *Asset) Deserialization(source *common.ZeroCopySource) error {
+	name, _, irregular, eof := source.NextString()
```
Comment (laizy) [inline core/asset/asset.go:113]: check `irregular` here

File: core/asset/config.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (laizy) [inline core/asset/config.json]: remove this file

File: core/code/FunctionCode.go [hunk 6]
```diff
@@ -51,6 +51,26 @@ func (fc *FunctionCode) Deserialize(r io.Reader) error {
...
 
+func (fc *FunctionCode) Serialization(sink *ZeroCopySink) error {
+	sink.WriteVarBytes(ContractParameterTypeToByte(fc.ParameterTypes))
+	sink.WriteVarBytes(fc.Code)
+	return nil
+}
+
+func (fc *FunctionCode) Deserialization(source *ZeroCopySource) error {
+	var eof, irregular bool
+	paramType, _, irregular, eof := source.NextVarBytes()
```
Comment (laizy) [inline core/code/FunctionCode.go]: check irregular here

File: core/contract/program/Program.go [hunk 7]
```diff
@@ -29,6 +30,28 @@ func (p *Program) Serialize(w io.Writer) error {
...
+func (p *Program) Serialization(sink *common.ZeroCopySink) error {
+	sink.WriteVarBytes(p.Parameter)
+	sink.WriteVarBytes(p.Code)
+	return nil
+}
+
+func (p *Program) Deserialization(source *common.ZeroCopySource) error {
+	var irregular, eof bool
+	var data []byte
+	data, _, irregular, eof = source.NextVarBytes()
```
Comment (laizy) [inline core/contract/program/Program.go:42]: check irregular

File: core/ledger/block.go [hunk 8]
```diff
@@ -37,6 +37,46 @@ func (b *Block) Serialize(w io.Writer) error {
...
 }
 
+func (b *Block) Serialization(sink *ZeroCopySink) error {
+	err := b.Blockdata.Serialization(sink)
+	if err != nil {
+		return err
+	}
+	sink.WriteUint32(uint32(len(b.Transactions)))
+	for _, transaction := range b.Transactions {
+		transaction.Serialization(sink)
```
Comment (laizy) [inline core/ledger/block.go]: handle error

File: core/ledger/block.go [hunk 9]
```diff
@@ -37,6 +37,46 @@ func (b *Block) Serialize(w io.Writer) error {
...
+}
+
+func (b *Block) Deserialization(source *ZeroCopySource) error {
+	if b.Blockdata == nil {
+		b.Blockdata = new(Blockdata)
+	}
+	b.Blockdata.Deserialization(source)
+	var eof bool
+    Len, eof := source.NextUint32()
+    if eof {
```
Comment (laizy) [inline core/ledger/block.go]: format code

File: core/ledger/block.go [hunk 10]
```diff
@@ -37,6 +37,46 @@ func (b *Block) Serialize(w io.Writer) error {
...
+    	return io.ErrUnexpectedEOF
+	}
+	var txhash Uint256
+	var tharray []Uint256
+	var i uint32
+	for i = 0; i < Len; i++ {
+		transaction := new(tx.Transaction)
+		transaction.Deserialization(source)
```
Comment (laizy) [inline core/ledger/block.go]: handle error

File: crypto/crypto.go [hunk 11]
```diff
@@ -127,6 +128,24 @@ func (e *PubKey) Serialize(w io.Writer) error {
 	return nil
 }
 
+func (e *PubKey) Serialization(sink *common.ZeroCopySink) error {
+	bufX := []byte{}
```
Comment (laizy) [inline crypto/crypto.go]: avoid duplicated code

**Review by laizy — CHANGES_REQUESTED** [review]

File: common/uint160.go (hunk 1 above)
Comment (lucas7788) [reply common/uint160.go]: ok

File: core/asset/asset.go [hunk 12]
```diff
@@ -65,6 +66,15 @@ func (a *Asset) Serialize(w io.Writer) error {
 	return nil
 }
 
+func (a *Asset)Serialization(sink *common.ZeroCopySink) error {
+	sink.WriteString(a.Name)
+	sink.WriteString(a.Description)
+	sink.WriteByte(byte(a.Precision))
+    sink.WriteByte(byte(a.AssetType))
```
Comment (laizy) [inline core/asset/asset.go]: format code

File: core/ledger/block.go [hunk 13]
```diff
@@ -37,6 +37,52 @@ func (b *Block) Serialize(w io.Writer) error {
...
+		}
+	}
+	return nil
+}
+
+func (b *Block) Deserialization(source *ZeroCopySource) error {
+	if b.Blockdata == nil {
+		b.Blockdata = new(Blockdata)
+	}
+	b.Blockdata.Deserialization(source)
```
Comment (laizy) [inline core/ledger/block.go]: check error

File: core/ledger/blockdata.go [hunk 14]
```diff
@@ -46,6 +55,17 @@ func (bd *Blockdata) SerializeUnsigned(w io.Writer) error {
...
 
+func (bd *Blockdata) SerializationUnsigned(sink *ZeroCopySink) error {
+    sink.WriteUint32(bd.Version)
+    sink.WriteUint256(bd.PrevBlockHash)
+    sink.WriteUint256(bd.TransactionsRoot)
+    sink.WriteUint32(bd.Timestamp)
+    sink.WriteUint32(bd.Height)
+    sink.WriteUint64(bd.ConsensusData)
+    sink.WriteUint160(bd.NextBookKeeper)
+	return nil
```
Comment (laizy) [inline core/ledger/blockdata.go:69]: format code

File: core/ledger/header.go [hunk 15]
```diff
@@ -15,6 +16,21 @@ func (h *Header) Serialize(w io.Writer) {
 
 }
 
+func (h *Header) Serialization(sink *common.ZeroCopySink) error {
+	h.Blockdata.Serialization(sink)
+	sink.WriteByte(byte(0))
```
Comment (laizy) [inline core/ledger/header.go]: `0` is not equal `'0'`

**Review by laizy — CHANGES_REQUESTED** [review]

**Review by laizy — APPROVED** [review]

**Pull request merged**